### PR TITLE
Improve bar_label annotation

### LINKED ISSUE #22414: [Bug]: bar_label overlaps bars when y-axis is inverted
### Bug summary

bar_label overlaps bars when the y-axis is inverted

### Code for reproduction

```python
import matplotlib.pyplot as plt
import numpy as np


labels = ['G1', 'G2', 'G3', 'G4', 'G5']
men_means = [20, 34, 30, 35, 27]
women_means = [25, 32, 34, 20, 25]

x = np.arange(len(labels))  # the label locations
width = 0.35  # the width of the bars

fig, ax = plt.subplots()
rects1 = ax.bar(x - width/2, men_means, width, label='Men')
rects2 = ax.bar(x + width/2, women_means, width, label='Women')

# Add some text for labels, title and custom x-axis tick labels, etc.
ax.set_ylabel('Scores')
ax.set_title('Scores by group and gender')
ax.legend()
ax.invert_yaxis()
ax.bar_label(rects1, padding=3)
ax.bar_label(rects2, padding=3)

fig.tight_layout()

plt.show()
```

### Actual outcome

Figure_1

### Expected outcome

The bar_label most be stay at the end of the bar without overlapping it, but with the y-axis inverted
Figure_1

### Additional information

_No response_

### Operating system

_No response_

### Matplotlib Version

3.5.1

### Matplotlib Backend

_No response_

### Python version

_No response_

### Jupyter version

_No response_

### Installation

pip

### PR DESCRIPTION
**Pull request by Valdes-Tresanco-MS**
## PR Summary
Improve bar_label method
Fix #22414
## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -2608,6 +2608,11 @@ def bar_label(self, container, labels=None, *, fmt="%g", label_type="edge",
 
         # want to know whether to put label on positive or negative direction
         # cannot use np.sign here because it will return 0 if x == 0
+        a, b = self.yaxis.get_view_interval()
+        yinverted = -1 if a > b else 1
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py:2612]: Why not use `True` / `False` here?  This looks like it is boolean state (and is unlikely to ever grow a third state).

**Comment by tacaswell** [comment]
Thank you for both reporting and fixing this @Valdes-Tresanco-MS !

I left one small comment in the code about the implementation.  Could you also add a test so that we are sure that it does not come back?  I do not think we need an image test.  I suspect there is a way to query the location of the text, convert it to data coordinates and assert that it is greater than the data it is above.

**Comment by Valdes-Tresanco-MS** [comment]
I described the solution in #22414 and in this comment https://github.com/matplotlib/matplotlib/issues/22414#issuecomment-1035402549. In this comment, I include also an image with the test result.

This is enough?
In the issue, @mwaskom also suggests a deeper revision, but unfortunately, I don't have time for it

Test code

```python
f, axes = plt.subplots(3, 4, figsize=(16, 12))
(ax1, ax2, ax3, ax4), (ax5, ax6, ax7, ax8), (ax9, ax10, ax11, ax12) = axes
bars1 = ax1.bar(["a", "b", "c", "d"], [1, 2, 3, -3])
ax1.bar_label(bars1, padding=0, label_type='center')
ax1.set_title('Vertical Normal yaxis padding=0')

bars2 = ax2.bar(["a", "b", "c", "d"], [1, 2, 3, -3])
ax2.invert_yaxis()
ax2.bar_label(bars2, padding=0, label_type='center')
ax2.set_title('Vertical Inverted yaxis padding=0')

bars5 = ax5.barh(["a", "b", "c", "d"], [1, 2, 3, -3])
ax5.bar_label(bars5, padding=0, label_type='center')
ax5.set_title('Horizontal Normal yaxis padding=0')

bars6 = ax6.barh(["a", "b", "c", "d"], [1, 2, 3, -3])
ax6.invert_yaxis()
ax6.bar_label(bars6, padding=0, label_type='center')
ax6.set_title('Horizontal Inverted yaxis padding=0')

bars9 = ax9.barh(["a", "b", "c", "d"], [1, 2, 3, -3])
ax9.invert_xaxis()
ax9.bar_label(bars9, padding=0, label_type='center')
ax9.set_title('Horizontal Inverted xaxis padding=0')

bars10 = ax10.barh(["a", "b", "c", "d"], [1, 2, 3, -3])
ax10.invert_yaxis()
ax10.invert_xaxis()
ax10.bar_label(bars10, padding=0, label_type='center')
ax10.set_title('Horizontal Inverted yaxis and xaxis padding=0')

bars3 = ax3.bar(["a", "b", "c", "d"], [1, 2, 3, -3])
ax3.bar_label(bars3, padding=10, label_type='center')
ax3.set_title('Vertical Normal yaxis padding=10')

bars4 = ax4.bar(["a", "b", "c", "d"], [1, 2, 3, -3])
ax4.invert_yaxis()
ax4.bar_label(bars4, padding=10, label_type='center')
ax4.set_title('Vertical Inverted yaxis padding=10')

bars7 = ax7.barh(["a", "b", "c", "d"], [1, 2, 3, -3])
ax7.bar_label(bars7, padding=10, label_type='center')
ax7.set_title('Horizontal Normal yaxis padding=10')

bars8 = ax8.barh(["a", "b", "c", "d"], [1, 2, 3, -3])
ax8.invert_yaxis()
ax8.bar_label(bars8, padding=10, label_type='center')
ax8.set_title('Horizontal Inverted yaxis padding=10')

bars11 = ax11.barh(["a", "b", "c", "d"], [1, 2, 3, -3])
ax11.invert_xaxis()
ax11.bar_label(bars11, padding=10, label_type='center')
ax11.set_title('Horizontal Inverted xaxis padding=10')

bars12 = ax12.barh(["a", "b", "c", "d"], [1, 2, 3, -3])
ax12.invert_yaxis()
ax12.invert_xaxis()
ax12.bar_label(bars12, padding=10, label_type='center')
ax12.set_title('Horizontal Inverted yaxis and xaxis padding=10')

f.tight_layout()
```

**Comment by mwaskom** [comment]
Worth pointing out that this is now inconsistent with how annotations work:

```python
ax = plt.gca()
ax.invert_yaxis()
ax.bar([0, 1], [1, .5])
ax.annotate("hello", (1, .5), xytext=(0, 10), textcoords="offset points", ha="center", va="center")
```
image

I'm not sure if there is other precedence for point offset adjustments in the library to cross-reference.

**Comment by QuLogic** [comment]
Please give your PRs and commits more descriptive titles; no-one will know what 22414 is in two weeks.

**Comment by Valdes-Tresanco-MS** [comment]
@mwaskom Sorry if I missed something, but why wouldn't it be inconsistent? `bar_label` makes an annotation passing modified arguments 
https://github.com/matplotlib/matplotlib/blob/541a7867ad35cad2c1d7f20905edbda7087a09a6/lib/matplotlib/axes/_axes.py#L2685-L2688

If I add this line to your example, everything seems to work fine. 
```python
ax = plt.gca()
ax.invert_yaxis()
b = ax.bar([0, 1], [1, .5])
ax.annotate("hello", (1, .5), xytext=(0, 20), textcoords="offset points", ha="center", va="center")
ax.bar_label(b)
```
output

Suddenly something related to the handling of the attributes of the annotations escapes me, so I would not like to be left with the doubt.

Additionally, could this be a solution to what you suggest?
https://github.com/matplotlib/matplotlib/blob/541a7867ad35cad2c1d7f20905edbda7087a09a6/lib/matplotlib/axes/_axes.py#L2680-L2688

```python
if 'ha' not in kwargs:
    kwargs.update([('ha', ha)])
if va not in kwargs:
    kwargs.update([('va', va)])

annotation = self.annotate(fmt % value if lbl is None else lbl,
                           xy, xytext, textcoords="offset points",
                           **kwargs)
```
Although it is not elegant in principle, it does what you suggest. If the alignment parameters are not explicitly defined, then those determined in the method are passed, otherwise, the user's parameters are used. I only tested a few cases, so I don't know if it could have a wrong behavior in some cases

**Comment by mwaskom** [comment]
> Sorry if I missed something, but why wouldn't it be inconsistent? bar_label makes an annotation passing modified arguments

Right, the existing behavior is happening because of the way that `annotate` works. I'm just pointing out that "fixing" it will introduce an inconsistency. I am not sure that I have an opinion about which behavior is preferable, just noting.

> could this be a solution to what you suggest?

Yes but it can be even simpler:

```python
kwargs.setdefault("ha", ha)
kwargs.setdefault("va", va)
```

**Comment by tacaswell** [comment]
@Valdes-Tresanco-MS Posts on GH are certainly useful!  However, the test code needs to end up in the test suite that is run on PR so the computer can tell us if things are broken.

**Comment by Valdes-Tresanco-MS** [comment]
> ```python
> kwargs.setdefault("ha", ha)
> kwargs.setdefault("va", va)
> ```

I found a small problem with the implementation of the `ha` and `va` parameters defined by the user. Since the annotation, in this case, is done on each bar, assigning the items `ha` and `va` envelopes overwrites the configuration, making all labels appear the same, whether they are for positive or negative values. The solution is to create a copy of `kwargs` on each iteration to define these particular variables in each iteration. However, the fact that the labels are defined in the same way for all the values I think can be confusing for users.

**Comment by Valdes-Tresanco-MS** [comment]
@tacaswell I think it's ready.

**Comment by Valdes-Tresanco-MS** [comment]
@tacaswell I see you added it to the `3.6.0` milestone. Since it is an improvement of a method that was already implemented, is it possible to add it in the current version (`3.5.x`)? It would be a great help for my projects.
Thanks

File: lib/matplotlib/axes/_axes.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py:2612]: This is rather a sign than a boolean. `+/-1` are the right values. The confusion comes from the fact that the variable name hints at a bool. One could use e.g. `y_direction`, but that'd need an explanation as well.

Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py:2612]: We have two usages below: a boolean and a sign. This tries to push both notions into one variable, which makes it a bit odd: The variable naming suggests a boolean, but the value carries the sign.

I suggest
```suggestion
        yinverted = a > b
```
and create the sign below; see there.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -2665,18 +2670,24 @@ def sign(x):
                 xy = endpt, yc
 
             if orientation == "vertical":
-                xytext = 0, sign(dat) * padding
+                xytext = 0, yinverted * sign(dat) * padding
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:2673]:
```suggestion
                ydirection = -1 if yinverted else 1
                xytext = 0, ydirection * sign(dat) * padding
```
Same for `xinverted` below

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -2665,18 +2670,24 @@ def sign(x):
...
-                xytext = sign(dat) * padding, 0
+                xytext = xinverted * sign(dat) * padding, 0
 
             if label_type == "center":
                 ha, va = "center", "center"
             elif label_type == "edge":
                 if orientation == "vertical":
                     ha = 'center'
-                    va = 'top' if dat < 0 else 'bottom'  # also handles NaN
+                    if yinverted == -1:
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:2682]: And then
```suggestion
                    if yinverted:
```

**Comment by Valdes-Tresanco-MS** [comment]
@timhoffm I see, definitely better

**Comment by timhoffm** [comment]
@Valdes-Tresanco-MS since you wrote that you don't have much time, I took the liberty and pushed my proposed changes as well as error handling if the user passes alignment parameters to your branch. I hope that's ok.

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -2598,16 +2598,28 @@ def bar_label(self, container, labels=None, *, fmt="%g", label_type="edge",
 
         **kwargs
             Any remaining keyword arguments are passed through to
-            `.Axes.annotate`.
+            `.Axes.annotate`. The alignment parameters (
+            *horizontalalignment* / *ha*, *verticalalignment* / *va*) are
+            not supported because the labels are automatically aligned to
+            the bars.
```
Comment (mwaskom) [inline lib/matplotlib/axes/_axes.py:2604]: This is a frustrating limitation of the functionality.

**Review by oscargus — APPROVED** [review]

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -2598,16 +2598,28 @@ def bar_label(self, container, labels=None, *, fmt="%g", label_type="edge",
...
         list of `.Text`
             A list of `.Text` instances for the labels.
         """
+        for key in ['horizontalalignment', 'ha', 'verticalalignment', 'va']:
+            if key in kwargs:
+                raise ValueError(
+                    f"Passing {key!r} to bar_label() is not supported.")
 
         # want to know whether to put label on positive or negative direction
         # cannot use np.sign here because it will return 0 if x == 0
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py:2617]: This comment is about the following function, and should not be split away from it.

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -2665,18 +2677,26 @@ def sign(x):
...
             if label_type == "center":
                 ha, va = "center", "center"
             elif label_type == "edge":
                 if orientation == "vertical":
                     ha = 'center'
-                    va = 'top' if dat < 0 else 'bottom'  # also handles NaN
+                    if y_inverted:
+                        va = 'top' if dat > 0 else 'bottom'  # also handles NaN
+                    else:
+                        va = 'top' if dat < 0 else 'bottom'  # also handles NaN
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py:2694]: I'm not entirely certain this is correct, because there are no tests with NaN on an inverted axis.

Comment (jklymak) [reply lib/matplotlib/axes/_axes.py:2694]: @Valdes-Tresanco-MS   di this get addressed?  If so, please move back out of draft.   Thanks!

**Comment by Valdes-Tresanco-MS** [comment]
@jklymak Sorry, I don't understand what I should do. Since it was indicated I included the test (a64db70). The failure in the check seems to be related to a CLI failure not to the test itself. This commit should I not include it?

**Comment by jklymak** [comment]
I just wanted to be sure @QuLogic 's comment had been addressed....

**Comment by Valdes-Tresanco-MS** [comment]
> I just wanted to be sure @QuLogic 's comment had been addressed....

Yes. Should I mark the conversations as resolved? It's my first PR here

File: lib/matplotlib/axes/_axes.py [hunk 7]
```diff
@@ -2598,13 +2598,25 @@ def bar_label(self, container, labels=None, *, fmt="%g", label_type="edge",
...
+            not supported because the labels are automatically aligned to
+            the bars.
 
         Returns
         -------
         list of `.Text`
             A list of `.Text` instances for the labels.
         """
+        for key in ['horizontalalignment', 'ha', 'verticalalignment', 'va']:
+            if key in kwargs:
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:2612]: Might alternatively written as
```suggestion
        unsupported = set(['horizontalalignment', 'ha', 'verticalalignment', 'va']):
        for key in unsupported.intersection(kwargs):
```
But no strong preference. I'll leave this up to you.

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
*Python 3.8 on ubuntu-18.04* test failure is unrelated.

File: lib/matplotlib/axes/_axes.py (hunk 7 above)
Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py:2612]: That should be:
```
unsupported = {'horizontalalignment', 'ha', 'verticalalignment', 'va'}
```

File: lib/matplotlib/axes/_axes.py (hunk 6 above)
Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py:2694]: There was a new test added, but the problem is that the result that is being checked is the same as what occurs when you run the code on `main`. So I'm not certain if the test is asserting that this PR changes things correctly (and the test just happens to match `main` coincidentally), or the test is asserting that this PR doesn't break how things currently work (and the match is intentional).

Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py:2694]: So what I guess I'm saying is it would be better if the test were paramterized across all the 4 cases here.

**Comment by timhoffm** [comment]
@Valdes-Tresanco-MS thanks, and congratulations on your first contribution to Matplotlib! We hope to see you back.

File: lib/matplotlib/axes/_axes.py (hunk 7 above)
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py:2612]: Merged as is since the existing code style is good enough.

**Comment by Valdes-Tresanco-MS** [comment]
> @Valdes-Tresanco-MS thanks, and congratulations on your first contribution to Matplotlib! We hope to see you back.

Definitely... I will try to contribute what I can in the future.

**Pull request merged**